doc/shared/sunadjoint/SUNAdjointStepper.rst

```diff
@@ -0,0 +1,247 @@
+..
+   ----------------------------------------------------------------
+   SUNDIALS Copyright Start
+   Copyright (c) 2002-2024, Lawrence Livermore National Security
+   and Southern Methodist University.
+   All rights reserved.
+
+   See the top-level LICENSE and NOTICE files for details.
+
+   SPDX-License-Identifier: BSD-3-Clause
+   SUNDIALS Copyright End
+   ----------------------------------------------------------------
+
+.. _SUNAdjointStepper:
+
+The SUNAdjointStepper API and Module
+====================================
+
+The :c:type:`SUNAdjointStepper` API and module provides a package-agnostic
+interface to SUNDIALS ASA capabilities. It currently only supports the discrete
+ASA capabilities in the ARKODE package, but in the future this support may be expanded.
+
+A :c:type:`SUNAdjointStepper` is a pointer to the
+:c:struct:`SUNAdjointStepper_` structure:
+
+.. c:type:: struct SUNAdjointStepper_ *SUNAdjointStepper
+
+.. c:struct:: SUNAdjointStepper_
+
+   .. c:member:: SUNStepper adj_sunstepper
+
+      The :c:type:`SUNStepper` object used for backwards time stepping of the adjoint ODE system.
+
+   .. c:member:: SUNStepper fwd_sunstepper
+
+      The :c:type:`SUNStepper` object used for forward time stepping of the original ODE system if any recomputation of missing
+      state data is required during the backwards integration.
+
+   .. c:member:: sunrealtype tf
+
+      The terminal time of the backwards adjoint ODE.
+
+   .. c:member:: int64_t step_idx
```

Comment:
I think we should be OK because the indexes are all based on ARKODE's step count. As such, the value should always fit in a `long int` and the (potential) downcast should not matter. Let me know if you see somewhere this is not true.